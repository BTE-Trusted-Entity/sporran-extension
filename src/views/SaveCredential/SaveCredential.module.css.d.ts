export const __esModule: true;
export const container: string;
export const heading: string;
export const cardContainer: string;
export const warning: string;
export const downloaded: string;
export const download: string;
export const done: string;
export const close: string;
