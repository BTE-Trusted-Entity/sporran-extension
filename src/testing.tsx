export * from '@testing-library/react';
export { waitForElementToBeRemoved } from '@testing-library/dom';
import { act, render as externalRender, screen } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import dialogPolyfill from 'dialog-polyfill';

import { AccountsProviderMock } from './testing/AccountsProviderMock';

export {
  AccountsProviderMock,
  accountsMock,
} from './testing/AccountsProviderMock';

export { mockBackgroundScript } from './testing/mockBackgroundScript';

export function render(
  ui: Parameters<typeof externalRender>[0],
  options?: Parameters<typeof externalRender>[1],
): ReturnType<typeof externalRender> {
  return externalRender(
    <AccountsProviderMock>
      <MemoryRouter>{ui}</MemoryRouter>
    </AccountsProviderMock>,
    options,
  );
}

export async function waitForTooltipUpdate(): Promise<void> {
  expect(await screen.findByRole('status')).toHaveAttribute(
    'data-popper-escaped',
  );
}

const dialogPromise = Promise.resolve();
(dialogPolyfill.registerDialog as jest.Mock).mockReturnValue(dialogPromise);

export async function waitForDialogUpdate(): Promise<void> {
  await act(() => dialogPromise);
}
