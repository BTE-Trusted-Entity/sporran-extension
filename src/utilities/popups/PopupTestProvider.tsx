import { MemoryRouter } from 'react-router-dom';

import { jsonToBase64 } from '../base64/base64';

interface Props {
  path: string;
  data: unknown;
  children: JSX.Element;
}

export function PopupTestProvider({
  path,
  data,
  children,
}: Props): JSX.Element {
  const serialized = jsonToBase64(data);
  return (
    <MemoryRouter initialEntries={[`${path}?data=${serialized}`]}>
      {children}
    </MemoryRouter>
  );
}
