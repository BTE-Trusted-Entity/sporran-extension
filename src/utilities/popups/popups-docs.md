# Integration of websites with the Sporran extension

We need to provide JS api for the websites to be able to interact with the extension.
Apparently the only way for an extension to do that is to inject a script into
the webpage which will create that API.

We also need to show some extension UI in the response to the API calls.
The browsers intentionally prevent JS code from showing the standard extension popup.
Luckily the extension has an internal URI and its UI can be shown in a normal browser
window (or popup window). However only the code running in the extension background
process or popup process can do this. The browsers intentionally prevent JS code
on websites, in injected scripts, as well as in the content scripts from accessing
the extension UI URIs.

Finally, we also need to pass the API call data to the extension UI.
The UI uses internally an in-memory router which is unaware about the actual URI.

This is how the system has to work in such environment:

1. The extension API is exposed to the website JS by an injected script.
   In response to the API calls this script passes messages to the content script.
1. The content script injects the injected script into the webpage.
   It also listens for the messages from it and passes them on to the background process.
1. The background process listens to the messages from the content script.
   When a message is received, it constructs an internal extension URI and
   encodes in it the parameters of the call. Then it opens this URI
   in a new browser window.
1. When the extension UI initializes it detects that it is shown in the browser window,
   not in the extension popup. In that case it takes the parameters encoded in the URI,
   and transforms them for use in the internal router.
