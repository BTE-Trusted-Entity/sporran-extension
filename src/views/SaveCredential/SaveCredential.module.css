.container {
  composes: container from '../../components/View/View.module.css';
}

.heading {
  composes: screenHeadline from '../../components/Typography/Typography.module.css';
  margin-bottom: 1rem;
}

.cardContainer {
  composes: fullWidthContainer from '../../components/View/View.module.css';
  list-style: none;
  max-height: calc(250rem / 16);
}

.warning {
  composes: importantHeadline from '../../components/Typography/Typography.module.css';
  margin: 1rem 0;
  font-size: 1.375rem;
}

.downloaded {
  composes: warning;
  color: var(--color-lightest);
}

.download {
  composes: buttonPrimary from '../../components/Button/Button.module.css';
}

.done {
  composes: subline from '../../components/Typography/Typography.module.css';
}

.close {
  composes: buttonSecondary from '../../components/Button/Button.module.css';
  padding: 0;
}
