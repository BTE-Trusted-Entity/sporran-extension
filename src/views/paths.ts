export { generatePath } from 'react-router-dom';

export const paths = {
  home: '/',
  settings: '/settings',
  access: '/access',
  identity: {
    base: '/identity/',
    add: '/identity/add',
    create: {
      start: '/identity/create',
      backup: '/identity/create/backup',
      verify: '/identity/create/verify',
      password: '/identity/create/password',
    },
    import: {
      start: '/identity/import',
      password: '/identity/import/password',
    },

    overview: '/identity/:address/:type(created|imported|pwreset)?',
    reset: {
      start: '/identity/:address/reset',
      password: '/identity/:address/reset/password',
    },
    send: {
      start: '/identity/:address/send',
      review: '/identity/:address/send/review',
      warning: '/identity/:address/send/warning',
    },
    receive: '/identity/:address/receive',
    credentials: {
      base: '/identity/:address/credentials',
      presentation: '/identity/:address/credentials/:hash/presentation',
    },
    remove: '/identity/:address/remove',
    did: {
      upgrade: {
        start: '/identity/:address/did/upgrade',
        kilt: '/identity/:address/did/upgrade/kilt',
        euro: '/identity/:address/did/upgrade/euro',
      },
      manage: {
        start: '/identity/:address/did',
        downgrade: {
          base: '/identity/:address/did/downgrade',
          warning: {
            web3name: '/identity/:address/did/downgrade/warning/web3name',
            credentials: '/identity/:address/did/downgrade/warning/credentials',
          },
          sign: '/identity/:address/did/downgrade/sign',
        },
        endpoints: {
          start: '/identity/:address/did/endpoints',
          edit: '/identity/:address/did/endpoints/:id?',
          sign: '/identity/:address/did/endpoints/sign',
        },
      },
      repair: '/identity/:address/did/repair',
    },
    web3name: {
      create: {
        base: '/identity/:address/web3name/create',
        info: '/identity/:address/web3name/create/info',
        form: '/identity/:address/web3name/create/form',
        choose: '/identity/:address/web3name/create/choose/:web3name',
        kilt: '/identity/:address/web3name/create/kilt/:web3name',
        euro: '/identity/:address/web3name/create/euro/:web3name',
      },
      manage: {
        start: '/identity/:address/web3name',
        remove: '/identity/:address/web3name/remove',
      },
    },
  },
  popup: {
    base: '/popup',
    import: '/popup/import',
    access: '/popup/access',
    claim: '/identity/:address/claim',
    save: '/popup/save',
    share: {
      start: '/popup/share',
      sign: '/popup/share/sign',
    },
    sign: '/popup/sign',
    signRaw: '/popup/signRaw',
    signDid: {
      start: '/identity/:address/did/sign',
      sign: '/identity/:address/did/sign/password',
      credentials: '/identity/:address/did/sign/credentials',
    },
    signDidExtrinsic: '/identity/:address/did/signExtrinsic',
    createDid: '/identity/:address/did/create',
  },
};
