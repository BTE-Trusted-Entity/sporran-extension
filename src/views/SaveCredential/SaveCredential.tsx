import { useCallback, useEffect } from 'react';
import { browser } from 'webextension-polyfill-ts';
import { IAttestation } from '@kiltprotocol/sdk-js';

import * as styles from './SaveCredential.module.css';

import {
  getCredentialDownload,
  saveCredential,
  useCredentials,
} from '../../utilities/credentials/credentials';
import { usePopupData } from '../../utilities/popups/usePopupData';

import { saveChannel } from '../../channels/saveChannel/saveChannel';

import { CredentialCard } from '../../components/CredentialCard/CredentialCard';

export function SaveCredential(): JSX.Element | null {
  const t = browser.i18n.getMessage;

  const { claimHash } = usePopupData<IAttestation>();

  const sporranCredentials = useCredentials();

  const sporranCredential = sporranCredentials?.find(
    ({ credential: { rootHash } }) => rootHash === claimHash,
  );

  const handleDownloadClick = useCallback(async () => {
    if (!sporranCredential) {
      return;
    }
    await saveCredential({ ...sporranCredential, isDownloaded: true });
  }, [sporranCredential]);

  const handleClose = useCallback(() => {
    window.close();
  }, []);

  useEffect(() => {
    (async () => {
      await saveChannel.return(undefined);
    })();
  }, []);

  if (!sporranCredentials) {
    return null; // storage data pending
  }

  if (!sporranCredential) {
    // TODO: decide on interface for an unknown credential
    return null;
  }

  const download = getCredentialDownload(sporranCredential);

  return (
    <main className={styles.container}>
      <h1 className={styles.heading}>{t('view_SaveCredential_heading')}</h1>

      <section className={styles.cardContainer}>
        <CredentialCard
          sporranCredential={sporranCredential}
          expand
          collapsible={false}
          buttons={false}
        />
      </section>

      <h2
        className={
          sporranCredential.isDownloaded ? styles.downloaded : styles.warning
        }
      >
        {t('view_SaveCredential_warning')}
      </h2>

      <a
        download={download.name}
        href={download.url}
        className={styles.download}
        onClick={handleDownloadClick}
      >
        {t('view_SaveCredential_CTA')}
      </a>

      {sporranCredential.isDownloaded && (
        <p className={styles.done}>
          {t('view_SaveCredential_done')}
          <button type="button" className={styles.close} onClick={handleClose}>
            {t('view_SaveCredential_close')}
          </button>
        </p>
      )}
    </main>
  );
}
