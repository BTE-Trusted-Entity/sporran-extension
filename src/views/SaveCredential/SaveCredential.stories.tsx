import { Meta } from '@storybook/react';

import { mockAttestations } from '../../utilities/credentials/CredentialsProvider.mock';
import { PopupTestProvider } from '../../utilities/popups/PopupTestProvider';
import { paths } from '../paths';

import { SaveCredential } from './SaveCredential';

export default {
  title: 'Views/SaveCredential',
  component: SaveCredential,
} as Meta;

const { downloaded, notDownloaded } = mockAttestations;

export function Downloaded(): JSX.Element {
  return (
    <PopupTestProvider path={paths.popup.save} data={downloaded}>
      <SaveCredential />
    </PopupTestProvider>
  );
}

export function NotDownloaded(): JSX.Element {
  return (
    <PopupTestProvider path={paths.popup.save} data={notDownloaded}>
      <SaveCredential />
    </PopupTestProvider>
  );
}
