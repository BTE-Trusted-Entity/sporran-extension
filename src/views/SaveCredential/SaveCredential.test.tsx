import { render, screen } from '../../testing/testing';
import { mockAttestations } from '../../utilities/credentials/CredentialsProvider.mock';
import { PopupTestProvider } from '../../utilities/popups/PopupTestProvider';
import { paths } from '../paths';

import { SaveCredential } from './SaveCredential';

describe('SaveCredential', () => {
  it('should render for not downloaded credential', async () => {
    const { container } = render(
      <PopupTestProvider
        path={paths.popup.save}
        data={mockAttestations.notDownloaded}
      >
        <SaveCredential />
      </PopupTestProvider>,
    );

    await screen.findByText('Trusted Entity attester');

    expect(container).toMatchSnapshot();
  });
  it('should render for downloaded credential', async () => {
    const { container } = render(
      <PopupTestProvider
        path={paths.popup.save}
        data={mockAttestations.downloaded}
      >
        <SaveCredential />
      </PopupTestProvider>,
    );

    await screen.findByText('Trusted Entity attester');

    expect(container).toMatchSnapshot();
  });
});
