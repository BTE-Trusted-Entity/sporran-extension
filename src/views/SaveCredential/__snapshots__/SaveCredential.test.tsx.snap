// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SaveCredential should render for downloaded credential 1`] = `
<div>
  <main
    class="container"
  >
    <h1
      class="heading"
    >
      Back Up Credential
    </h1>
    <section
      class="cardContainer"
    >
      <li
        aria-expanded="true"
        class="credential"
      >
        <section
          class="expanded"
        >
          <section
            class="buttons"
          />
          <dl
            class="details"
          >
            <div
              class="detail"
            >
              <dt
                class="detailName"
              >
                Credential name
              </dt>
              <dd
                class="nameValue"
              >
                Email Credential
                <button
                  aria-label="Edit credential name"
                  class="editName"
                />
              </dd>
            </div>
            <div
              class="detail"
            >
              <dt
                class="detailName"
              >
                Status
              </dt>
              <dd
                class="detailValue"
              >
                Attested
              </dd>
            </div>
            <div
              class="detail"
            >
              <dt
                class="detailName"
              >
                Email
              </dt>
              <dd
                class="detailValue"
              >
                mockEmail@mock.mock
              </dd>
            </div>
          </dl>
          <h4
            class="technical"
          >
            Technical details
          </h4>
          <dl
            class="details"
          >
            <div
              class="detail"
            >
              <dt
                class="detailName"
              >
                CType
              </dt>
              <dd
                class="detailValue"
              >
                Email
              </dd>
            </div>
            <div
              class="detail"
            >
              <dt
                class="detailName"
              >
                Attester
              </dt>
              <dd
                class="detailValue"
              >
                Trusted Entity attester
              </dd>
            </div>
            <div
              class="hash"
            >
              <dt
                class="detailName"
              >
                Credential hash
              </dt>
              <dd
                class="detailValue"
              >
                0xbe7fcc7aa0186b05fd116c100e2d673fb951163693788640ce6032ad2f700dae
              </dd>
            </div>
          </dl>
        </section>
        <div />
      </li>
    </section>
    <h2
      class="downloaded"
    >
      Please back up your credential now!
    </h2>
    <a
      class="download"
      download="Email Credential-Email.json"
      href="data:text/json;base64,eyJjcmVkZW50aWFsIjp7ImNsYWltIjp7ImNUeXBlSGFzaCI6IjB4MzI5MWJiMTI2ZTMzYjQ4NjJkNDIxYmZhYTFkMmYyNzJlNmNkZmM0Zjk2NjU4OTg4ZmJjZmZlYTg5MTRiZDlhYyIsImNvbnRlbnRzIjp7IkVtYWlsIjoibW9ja0VtYWlsQG1vY2subW9jayJ9LCJvd25lciI6ImRpZDpraWx0OmxpZ2h0OjAwNG9VaUszRXZUY3pjNnVrWUxFVjU3QnV4emtXOXprc0tSY2haWFlHU29jTG1XRTVOIn0sImNsYWltSGFzaGVzIjpbIjB4MTQ0NTk3ZDI4NDVlMzI1ZGZmZDRlNTFhOTRlM2U5YzA0YjUyZGMxNTczZTY2YWZjNjkzODQ2ZDkxZjMxYTcxMyIsIjB4MzE3N2VhNDFmNzZkMDY2ZDE5MTIzYzJjYjBkZTEzYTM3ZDYzOTA2ZDJjNTE5ZTM2ZmJmZDc3NzNmYWMxYjcxOCIsIjB4OWQyZGNhZjUyMzhkNmJjNGJjZjAzMjhhZWRiMTY4OTQzNmUwNzdlYmY2MTk4YjZjNGU5NjE2NjQ4MDdlYWVkZiIsIjB4Y2VlMTkzZmZkZmE2MzQ4NzkwN2RmZTA4NDhhZTE1MGQ0YTcxOTZjYzFlM2Q1YmQyYzg5YmVjYjU0MDJlZmMwNyJdLCJjbGFpbU5vbmNlTWFwIjp7IjB4NTc0NThhNjk3MmU3ODIyM2NkNGY3ZjRjNTkyMzZlYTc2ZjM4N2FlNTcxYWI3ZTc5NDYwZjU2ZjJhYTk3NjIzZiI6ImQwYzlkNjQ3LTQyZWYtNGQ0NC1iNjAzLTNiNzc2MjYwYTZkZCIsIjB4Y2VmZTJiODhkMTA4MzQ4NjlkYWQzZThkNzMwNmY1YWE5OWExZjA3ODIxNGFkZWFlNzljYmVlZDMwM2U2MzhhZiI6IjZiNjIzMDhjLTg1NTctNGI2MS1hYzAyLThmOTA1NTU1ZDY3YiIsIjB4OGJlMzJkN2U5Y2M1MDE1YmY3MWI1YzQ1NTUwYTg3OTBjNDNiOWEyNmRiZmE5YTg1MjM4NzEyMTE3NDVkMzNhMyI6IjI0NzM5NGMwLWQ0NmItNDgyYS1iMGM1LTQ5M2FlYjUwNmQ2MSIsIjB4MTM2MWE3YmE3NTEyNTZlOWFjNmIzZTNlMjQ5MTJlNTBlNjhiZWY2NzhjNTEzMmMxMWM5NTY3ODIyMGUzYmY1ZSI6IjA3Yjk1ZjE3LTNmMjMtNGE5ZS04ZjQzLTMzODgxMjczNjgyYyJ9LCJsZWdpdGltYXRpb25zIjpbXSwiZGVsZWdhdGlvbklkIjpudWxsLCJyb290SGFzaCI6IjB4YmU3ZmNjN2FhMDE4NmIwNWZkMTE2YzEwMGUyZDY3M2ZiOTUxMTYzNjkzNzg4NjQwY2U2MDMyYWQyZjcwMGRhZSJ9LCJuYW1lIjoiRW1haWwgQ3JlZGVudGlhbCIsImNUeXBlVGl0bGUiOiJFbWFpbCIsImF0dGVzdGVyIjoiVHJ1c3RlZCBFbnRpdHkgYXR0ZXN0ZXIiLCJzdGF0dXMiOiJhdHRlc3RlZCIsImlzRG93bmxvYWRlZCI6dHJ1ZX0="
    >
      Back Up Credential
    </a>
    <p
      class="done"
    >
      All done! You can 
      <button
        class="close"
        type="button"
      >
        close this popup now
      </button>
    </p>
  </main>
</div>
`;

exports[`SaveCredential should render for not downloaded credential 1`] = `
<div>
  <main
    class="container"
  >
    <h1
      class="heading"
    >
      Back Up Credential
    </h1>
    <section
      class="cardContainer"
    >
      <li
        aria-expanded="true"
        class="credential"
      >
        <section
          class="expanded"
        >
          <section
            class="buttons"
          />
          <dl
            class="details"
          >
            <div
              class="detail"
            >
              <dt
                class="detailName"
              >
                Credential name
              </dt>
              <dd
                class="nameValue"
              >
                Not downloaded Email Credential
                <button
                  aria-label="Edit credential name"
                  class="editName"
                />
              </dd>
            </div>
            <div
              class="detail"
            >
              <dt
                class="detailName"
              >
                Status
              </dt>
              <dd
                class="detailValue"
              >
                Attested
              </dd>
            </div>
            <div
              class="detail"
            >
              <dt
                class="detailName"
              >
                Email
              </dt>
              <dd
                class="detailValue"
              >
                mockEmail@mock.mock
              </dd>
            </div>
          </dl>
          <h4
            class="technical"
          >
            Technical details
          </h4>
          <dl
            class="details"
          >
            <div
              class="detail"
            >
              <dt
                class="detailName"
              >
                CType
              </dt>
              <dd
                class="detailValue"
              >
                Email
              </dd>
            </div>
            <div
              class="detail"
            >
              <dt
                class="detailName"
              >
                Attester
              </dt>
              <dd
                class="detailValue"
              >
                Trusted Entity attester
              </dd>
            </div>
            <div
              class="hash"
            >
              <dt
                class="detailName"
              >
                Credential hash
              </dt>
              <dd
                class="detailValue"
              >
                0x86a20051e6d1eecb0dd6e9c7b3b0cf946867e662b525ef2e9afe8428f6616ae0
              </dd>
            </div>
          </dl>
        </section>
        <div />
      </li>
    </section>
    <h2
      class="warning"
    >
      Please back up your credential now!
    </h2>
    <a
      class="download"
      download="Not downloaded Email Credential-Email.json"
      href="data:text/json;base64,eyJjcmVkZW50aWFsIjp7ImNsYWltIjp7ImNUeXBlSGFzaCI6IjB4MzI5MWJiMTI2ZTMzYjQ4NjJkNDIxYmZhYTFkMmYyNzJlNmNkZmM0Zjk2NjU4OTg4ZmJjZmZlYTg5MTRiZDlhYyIsImNvbnRlbnRzIjp7IkVtYWlsIjoibW9ja0VtYWlsQG1vY2subW9jayJ9LCJvd25lciI6ImRpZDpraWx0OjRvZUo3Nmhkajg0eG53Q05xaWpVSFVDVG1md1hnU1o0dm14TEVpVEVZZ1FkQkNjWiJ9LCJjbGFpbUhhc2hlcyI6WyIweDE0NDU5N2QyODQ1ZTMyNWRmZmQ0ZTUxYTk0ZTNlOWMwNGI1MmRjMTU3M2U2NmFmYzY5Mzg0NmQ5MWYzMWE3MTMiLCIweDMxNzdlYTQxZjc2ZDA2NmQxOTEyM2MyY2IwZGUxM2EzN2Q2MzkwNmQyYzUxOWUzNmZiZmQ3NzczZmFjMWI3MTgiLCIweDlkMmRjYWY1MjM4ZDZiYzRiY2YwMzI4YWVkYjE2ODk0MzZlMDc3ZWJmNjE5OGI2YzRlOTYxNjY0ODA3ZWFlZGYiLCIweGNlZTE5M2ZmZGZhNjM0ODc5MDdkZmUwODQ4YWUxNTBkNGE3MTk2Y2MxZTNkNWJkMmM4OWJlY2I1NDAyZWZjMDciXSwiY2xhaW1Ob25jZU1hcCI6eyIweDU3NDU4YTY5NzJlNzgyMjNjZDRmN2Y0YzU5MjM2ZWE3NmYzODdhZTU3MWFiN2U3OTQ2MGY1NmYyYWE5NzYyM2YiOiJkMGM5ZDY0Ny00MmVmLTRkNDQtYjYwMy0zYjc3NjI2MGE2ZGQiLCIweGNlZmUyYjg4ZDEwODM0ODY5ZGFkM2U4ZDczMDZmNWFhOTlhMWYwNzgyMTRhZGVhZTc5Y2JlZWQzMDNlNjM4YWYiOiI2YjYyMzA4Yy04NTU3LTRiNjEtYWMwMi04ZjkwNTU1NWQ2N2IiLCIweDhiZTMyZDdlOWNjNTAxNWJmNzFiNWM0NTU1MGE4NzkwYzQzYjlhMjZkYmZhOWE4NTIzODcxMjExNzQ1ZDMzYTMiOiIyNDczOTRjMC1kNDZiLTQ4MmEtYjBjNS00OTNhZWI1MDZkNjEiLCIweDEzNjFhN2JhNzUxMjU2ZTlhYzZiM2UzZTI0OTEyZTUwZTY4YmVmNjc4YzUxMzJjMTFjOTU2NzgyMjBlM2JmNWUiOiIwN2I5NWYxNy0zZjIzLTRhOWUtOGY0My0zMzg4MTI3MzY4MmMifSwibGVnaXRpbWF0aW9ucyI6W10sImRlbGVnYXRpb25JZCI6bnVsbCwicm9vdEhhc2giOiIweDg2YTIwMDUxZTZkMWVlY2IwZGQ2ZTljN2IzYjBjZjk0Njg2N2U2NjJiNTI1ZWYyZTlhZmU4NDI4ZjY2MTZhZTAifSwibmFtZSI6Ik5vdCBkb3dubG9hZGVkIEVtYWlsIENyZWRlbnRpYWwiLCJjVHlwZVRpdGxlIjoiRW1haWwiLCJhdHRlc3RlciI6IlRydXN0ZWQgRW50aXR5IGF0dGVzdGVyIiwic3RhdHVzIjoiYXR0ZXN0ZWQiLCJpc0Rvd25sb2FkZWQiOmZhbHNlfQ=="
    >
      Back Up Credential
    </a>
  </main>
</div>
`;
